### PR TITLE
ISSUE: 42 and 30, implement headers to catalog v1 and get MimeType

### LINKED ISSUE #42: MIME Types
Requires #30 

- [ ] MIME type detection
- [ ] GET with MIME type response headers

### PR DESCRIPTION
**Pull request by rustchain64**
Headers include metadata formatted as CBOR data, and 
Endpoints provided with a JSON decoded format.

Closes #42 #30

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -13,22 +13,34 @@ rayon = ["blake3/rayon"]
 
 [dependencies]
 anyhow = "1.0.69"
-axum = "0.6.9"
+#axum = "0.6.9"
+axum = { version = "0.6.9", features = ["headers"] }
```
Comment (cryptoquick) [inline Cargo.toml]: Why was this added? I'm not sure this needs to be added.

File: Cargo.toml [hunk 2]
```diff
@@ -13,22 +13,34 @@ rayon = ["blake3/rayon"]
...
+
 axum-macros = "0.3.4"
 bao = "0.12.1"
 blake3 = "1.3.3"
 bytes = "1.4.0"
-carbonado = "0.3.0-rc.8"
+
+#carbonado = "0.3.0-rc.8"
+
+carbonado = { path = "/Users/wesleyparr/Projects/carbonado/carbonado" }
```
Comment (cryptoquick) [inline Cargo.toml]: Not good

File: Cargo.toml [hunk 3]
```diff
@@ -13,22 +13,34 @@ rayon = ["blake3/rayon"]
...
+
 chrono = "0.4.23"
 clap = { version = "4.1.8", features = ["derive"] }
 directories = "5.0.1"
 flexi_logger = { version = "0.25.2", features = ["colors"] }
 flexi_syslog = "0.5.2"
+futures = "0.3.28"
```
Comment (cryptoquick) [inline Cargo.toml]: We shouldn't need this except for very specific cases, try to find a way to work around it

File: Cargo.toml [hunk 4]
```diff
@@ -13,22 +13,34 @@ rayon = ["blake3/rayon"]
...
 futures-util = "0.3.28"
 hex = "0.4.3"
+hyper = "0.14"
```
Comment (cryptoquick) [inline Cargo.toml]: This is nonsense, do not add this, we already have axum

File: Cargo.toml [hunk 5]
```diff
@@ -13,22 +13,34 @@ rayon = ["blake3/rayon"]
...
 # human_bytes = "0.4.1"
 # infer = "0.13.0"
+infer = { version = "0.15.0", default-features = false }
```
Comment (cryptoquick) [inline Cargo.toml]: Why did you turn off default features? The ones provided seem pretty useful.

File: Cargo.toml [hunk 6]
```diff
@@ -13,22 +13,34 @@ rayon = ["blake3/rayon"]
...
+lazy_static = "1.4.0"
```
Comment (cryptoquick) [inline Cargo.toml]: We actually use once_cell, we don't use lazy_static. I'm also skeptical this is needed to implement this feature.

File: src/backend/fs.rs [hunk 7]
```diff
@@ -68,6 +88,8 @@ pub async fn write_file<'a>(
         .try_collect()
         .await?;
 
+    let mime_type_value = shared_mime_type.lock().await;
```
Comment (cryptoquick) [inline src/backend/fs.rs]: This is a terrible approach. Please refactor to not use locks and instead just get the mimetype before segmentation. This might require doing that in the http server stream loop.

File: src/frontend/http.rs [hunk 8]
```diff
@@ -31,6 +32,9 @@ async fn write_file_handler(pk: &str, body: BodyStream, name: Option<String>) ->
             while remainder.len() < SEGMENT_SIZE {
                 if let Some(chunk) = body_stream.next().await {
                     let bytes = chunk?;
+
+                    let convert = &bytes as &[u8];
+                    trace!("Show me the convert: {:?}", convert);
```
Comment (cryptoquick) [inline src/frontend/http.rs]: Remove this

File: src/frontend/http.rs [hunk 9]
```diff
@@ -81,6 +86,7 @@ async fn post_file_named(
     }
 
     let hash = write_file_handler(&pk, body, Some(name)).await?;
+    //let buf = [0xFF, 0xD8, 0xFF, 0xAA];
```
Comment (cryptoquick) [inline src/frontend/http.rs]: Clean this up please

File: src/frontend/http.rs [hunk 10]
```diff
@@ -57,6 +61,7 @@ async fn write_file_handler(pk: &str, body: BodyStream, name: Option<String>) ->
 #[axum_macros::debug_handler]
 async fn post_file(
     Path(pk): Path<String>,
+
```
Comment (cryptoquick) [inline src/frontend/http.rs]: Clean this up, please

File: src/header.rs [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    pub fn new(metadata: Vec<u8>) -> Result<Self, Error> {
+        //type Error = CarbonadoError;
+        let cbor_data = serde_cbor::to_vec(&metadata)?;
+        println!("cbor data; {:?}", cbor_data);
+
+        let cbor_len = cbor_data.len() as u8;
+
+        println!("cbor length ; {:?}", cbor_len);
+
+        let magic_no = carbonado::constants::MAGICNO;
```
Comment (cryptoquick) [inline src/header.rs]: Cat files will use their own magic number, remember what I said:

Use this for magic no:
https://github.com/diba-io/carbonado/blob/518188593fc2a86bf2fa687f9165b83904c0d64f/src/constants.rs#L5
Except use:
`CARBONCAT00` instead of `CARBONADO00`

File: src/header.rs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                Some(metadata)
+            } else {
+                None
+            },
+        })
+    }
+
+    /// Creates a header to be prepended to files.
+    pub fn try_to_vec(&self) -> Result<Vec<u8>, CarbonadoError> {
+        let mut cbore_bytes = self.cbor_len.to_le_bytes().to_vec(); // 2 bytes
```
Comment (cryptoquick) [inline src/header.rs]: This variable is misspelled

File: tests/file.rs [hunk 13]
```diff
@@ -87,48 +91,23 @@ async fn read_write_delete_file() -> Result<()> {
...
-    let file_did_write = write_file(&Secp256k1PubKey(pk), file_stream, None)
-        .await
-        .is_ok();
-
-    if file_did_write {
-        info!(
-            "Write File Group One to Delete File as blake3_hash:: {} ",
-            file_did_write.to_string()
-        );
-    }
```
Comment (cryptoquick) [inline tests/file.rs]: This whole test is removed! Please put it back.

File: tests/file.rs [hunk 14]
```diff
@@ -87,48 +91,23 @@ async fn read_write_delete_file() -> Result<()> {
...
     info!("Write Delete:: Writing file if not exists in order to test delete");
-    let blake3_hash = write_file(&Secp256k1PubKey(pk), file_stream, None)
-        .await
-        .is_ok();
 
-    if blake3_hash {
-        info!(
-            "Write File in Group Two to Delete File as blake3_hash:: {} ",
-            blake3_hash.to_string()
-        );
```
Comment (cryptoquick) [inline tests/file.rs]: Put these back, please

File: src/structs.rs [hunk 15]
```diff
@@ -121,3 +123,13 @@ pub enum PubKey {
...
     Secp256k1(secp256k1::PublicKey),
 }
+
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+
+pub struct CborData {
+    /// Number of bytes added to pad to align zfec chunks and bao slices.
+    pub str_name: String,
+    pub str_date: String,
+    pub str_mime_type: String,
```
Comment (cryptoquick) [inline src/structs.rs]: Let's remove the str prefix:

```suggestion
    pub name: String,
    pub date: String,
    pub mime_type: String,
```

**Review by cryptoquick — CHANGES_REQUESTED** [review]
This is a good start, but there's definitely a lot of room for improvement.

**Comment by rustchain64** [comment]
Thanks for the encouragement  and input

**Comment by rustchain64** [comment]
WARNING; adding header to the file, breaks the test. 
File no longer matches origonal hash
> 202 )  trace!("Write file header, should be added");
> 203 ) // file.write_all(&cat_header_bytes)?;

**Comment by rustchain64** [comment]
In "[pass mime_type to tuple works](https://github.com/diba-io/carbonado-node/pull/52/commits/b0db5578af01e7f09cbb92b5216b71639f2c06b5)"

We have verified that the mime_type value passes the mime_type to the tuple and is to be parsed from the segment, within "fs.rs"

File: src/frontend/http.rs [hunk 16]
```diff
@@ -25,18 +27,43 @@ use crate::{
...
+                    // is_mime_type = "test_mime_type".to_string();
+
+                    let kind = infer::get(&buffer); // Use the extended buffer here
+                    match kind {
+                        Some(mime_type) => {
+                            is_mime_type = mime_type.mime_type().to_string();
+                        }
+                        None => debug!("no mime_type found"),
+                    };
```
Comment (cryptoquick) [inline src/frontend/http.rs]:
```suggestion
                    let mime_type = match kind {
                        Some(mime_type) => {
                            mime_type.mime_type().to_string()
                        }
                        None => "application/octet-stream".to_owned()
                    };
```

File: src/frontend/http.rs [hunk 17]
```diff
@@ -25,18 +27,43 @@ use crate::{
...
+
+                    debug!(">>>>>>>>>>>> PASS TO TUPLE MIME_TYPE {}", &is_mime_type);
```
Comment (cryptoquick) [inline src/frontend/http.rs]: Clean this up
```suggestion
```

File: src/frontend/http.rs [hunk 18]
```diff
@@ -25,18 +27,43 @@ use crate::{
...
+        ),
+        |(mut body_stream, mut remainder, mut is_mime_type)| async move {
             while remainder.len() < SEGMENT_SIZE {
                 if let Some(chunk) = body_stream.next().await {
                     let bytes = chunk?;
+
+                    let mut buffer = Vec::new();
+                    buffer.extend_from_slice(&bytes);
+
+                    // is_mime_type = "test_mime_type".to_string();
```
Comment (cryptoquick) [inline src/frontend/http.rs]:
```suggestion
```

File: src/frontend/http.rs [hunk 19]
```diff
@@ -25,18 +27,43 @@ use crate::{
 async fn write_file_handler(pk: &str, body: BodyStream, name: Option<String>) -> Result<String> {
     let pk = &Secp256k1PubKey::try_from(pk)?;
 
+    let is_mime_type = "default_mime_type".to_string();
```
Comment (cryptoquick) [inline src/frontend/http.rs]:
```suggestion
```

File: src/frontend/http.rs [hunk 20]
```diff
@@ -49,14 +76,15 @@ async fn write_file_handler(pk: &str, body: BodyStream, name: Option<String>) ->
     )
     .boxed();
 
-    let Blake3Hash(hash) = write_file(pk, file_stream, name).await?;
+    let Blake3Hash(hash) = write_file(pk, file_stream, name, &is_mime_type).await?;
```
Comment (cryptoquick) [inline src/frontend/http.rs]: Notice, the mime-type isn't being updated here... I don't think the async code has run yet.

File: src/frontend/http.rs [hunk 21]
```diff
@@ -25,18 +27,43 @@ use crate::{
...
+                    debug!(">>>>>>>>>>>> PASS TO TUPLE MIME_TYPE {}", &is_mime_type);
+
                     remainder.extend(bytes);
 
                     if remainder.len() >= SEGMENT_SIZE {
                         let segment = remainder.split_to(SEGMENT_SIZE);
                         trace!("Stream 1MB segment");
-                        return Ok(Some((segment.freeze(), (body_stream, remainder))));
+                        return Ok(Some((
+                            segment.freeze(),
```
Comment (cryptoquick) [inline src/frontend/http.rs]:
```suggestion
                            (segment.freeze(), mime_type)
```
This may be enough to pass the mime_type into fs using the FileStream

File: src/backend/fs.rs [hunk 22]
```diff
@@ -31,6 +33,7 @@ pub async fn write_file<'a>(
     pk: &Secp256k1PubKey,
     file_stream: FileStream,
     name: Option<String>,
+    mime_type: &str,
```
Comment (cryptoquick) [inline src/backend/fs.rs]:
```suggestion
```
Let's not pass it in here, but instead, set it from the new `(Bytes, String)` tuple here:
https://github.com/diba-io/carbonado-node/pull/52/files#diff-d731782a4b2603919d38af2315cedf79f4d9ed9e5d1b4a1c8b23d74a75f337e8R57

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Really close! These changes should help get this on the right track.

File: Cargo.toml [hunk 23]
```diff
@@ -3,6 +3,9 @@ name = "carbonado-node"
 version = "0.1.0"
 edition = "2021"
 
+[profile.release]
+debug = true
```
Comment (cryptoquick) [inline Cargo.toml]: Let's remove this

File: src/frontend/http.rs [hunk 24]
```diff
@@ -10,77 +16,195 @@ use axum::{
...
+                // Here we check if the inferred type is `text/xml`, which could be an SVG.
+                if calculated_mime_type == "text/xml" {
+                    // Check for the presence of the `<svg` tag in the buffer
+                    // We only read up to a certain number of bytes to check for SVG to prevent having to load a very large file into memory
+                    let svg_check_len = std::cmp::min(buffer.len(), 1024); // Check first 1024 bytes
+                    let content = std::str::from_utf8(&buffer[..svg_check_len]).unwrap_or("");
+                    if content.contains("<svg") {
+                        debug!(">>>>>>>>>>>>>>>> content.contains SVG ");
+
+                        calculated_mime_type = "image/svg+xml".to_string();
```
Comment (cryptoquick) [inline src/frontend/http.rs]: This is neat, but it's still not following my recommendation to use the path extension to detect the type.

**Review by cryptoquick — CHANGES_REQUESTED** [review]
Please clean up debug statements and commented code

**Review by cryptoquick — APPROVED** [review]
Looking really good. LGTM.
